crates/solidity/outputs/cargo/crate/src/backend/ir/ast/nodes.rs

```diff
@@ -0,0 +1,229 @@
+        assert!(ir_node.kind == TerminalKind::Identifier);
+        Self {
+            ir_node: Rc::clone(ir_node),
+            semantic: Rc::clone(semantic),
+        }
+    }
+
+    pub fn unparse(&self) -> String {
+        self.ir_node.unparse()
+    }
+
+    pub fn resolve_to_definition(&self) -> Option<Definition> {
+        let reference = self
+            .semantic
+            .binder()
+            .find_reference_by_identifier_node_id(self.ir_node.id())?;
+        let definition_id = reference.resolution.as_definition_id()?;
+        Some(Rc::new(DefinitionStruct::create(
+            definition_id,
+            &self.semantic,
+        )))
+    }
+
+    pub fn is_definition(&self) -> bool {
+        self.semantic
+            .binder()
+            .find_definition_by_identifier_node_id(self.ir_node.id())
+            .is_some()
+    }
+
+    pub fn is_reference(&self) -> bool {
+        self.semantic
+            .binder()
+            .find_reference_by_identifier_node_id(self.ir_node.id())
+            .is_some()
+    }
+}
+
+pub type YulIdentifierStruct = IdentifierStruct;
+pub type YulIdentifier = Rc<YulIdentifierStruct>;
+
+pub type Definition = Rc<DefinitionStruct>;
+
+pub struct DefinitionStruct {
+    definition_id: NodeId,
+    semantic: Rc<SemanticAnalysis>,
+}
+
+macro_rules! define_variant {
+    ($variant:ident) => {
+        paste! {
+            pub fn [<is_ $variant:snake:lower>](&self) -> bool {
+                matches!(self.internal_definition(), binder::Definition::$variant(_))
+            }
+
+            pub fn [<as_ $variant:snake:lower>](&self) -> Option<[<$variant Definition>]> {
+                if let binder::Definition::$variant(variant) = self.internal_definition() {
+                    Some(Rc::new([<$variant DefinitionStruct>]::create(&variant.ir_node, &self.semantic)))
+                } else {
+                    None
+                }
+            }
+        }
+    };
+    ($variant:ident, $deftype:ident) => {
+        paste! {
+            pub fn [<is_ $variant:snake:lower>](&self) -> bool {
+                matches!(self.internal_definition(), binder::Definition::$variant(_))
+            }
+
+            pub fn [<as_ $variant:snake:lower>](&self) -> Option<$deftype> {
+                if let binder::Definition::$variant(variant) = self.internal_definition() {
+                    Some(Rc::new([<$deftype Struct>]::create(&variant.ir_node, &self.semantic)))
+                } else {
+                    None
+                }
+            }
+        }
+    };
+}
+
+impl DefinitionStruct {
+    pub(crate) fn create(definition_id: NodeId, semantic: &Rc<SemanticAnalysis>) -> Self {
+        assert!(semantic
+            .binder()
+            .find_definition_by_id(definition_id)
+            .is_some());
+        Self {
+            definition_id,
+            semantic: Rc::clone(semantic),
+        }
+    }
+
+    fn internal_definition(&self) -> &binder::Definition {
```

Comment:
I see this PR adds definitions, but not references. Is this planned for a future PR?

Reply:
References would need to be added to the the `Definition` (or each `*Definition` AST node). It's not added yet because we don't have the inverse mapping computed yet on the `SemanticAnalysis` object, but that's easy to do. I think that can go into this PR. I'll add that.